### ISSUE TITLE
New Parameter Items: FilePicker, ActionWithShortcut, PopupLineEditor

### ISSUE THREAD

**Issue by ntjess** [pyqtgraph/pyqtgraph#1378]
I ended up making these for one of my application and am proposing them here for integration into pyqtgraph, provided they are general enough. What do you think? Is it worth the effort of a pull request / should I post code in this issue / something else?

* The file dialogs are, of course, native to the OS since it is a `QFileDialog`.
* Folder and file pickers are the same parameter, where a boolean option `asFolder` determines whether files or folders are selected
* Similar to `list`, the options for `popuplineeditor` are set in the `limits` parameter. A `QListModel` with a `completer` are the core elements.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/23620506/93737986-9de4f600-fbb2-11ea-8239-b365a12d25f6.gif)

**Comment by ntjess** [comment]
FYI the source code for these parameter types is located in `pgregistered.py` within s3a ([here](https://gitlab.com/ficsresearch/s3a/-/blob/development/s3a/parameditors/pgregistered.py))

**Comment by ntjess** [comment]
Also introducing `SliderParameter`:

```python
children = [
  dict(name='a', type='slider', value=5, limits=[0, 10]),
  dict(name='b', type='slider', value=0.1, limits=[-5, 5], step=0.1),
  dict(name='c', type='slider', value=2, span=np.linspace(0, 2*np.pi, 1000)),
]

p = Parameter.create(name='test', type='group', children=children)
t = ParameterTree()
t.setParameters(p, showTop=True)
```
![HUFpoYVAhN](https://user-images.githubusercontent.com/23620506/99867653-3125b200-2b89-11eb-8141-9a5f590545f0.gif)

**Comment by ntjess** [comment]
@j9ac9k should I create a pull request with these? Or do you anticipate it will be rejected until usage is standardized / agreed upon?

**Comment by ntjess** [comment]
@j9ac9k I see there was some recent attention on this ticket. Just for reference, the implementation for all these parameters is now [here](https://gitlab.com/ntjess/utilitys/-/blob/master/utilitys/params/pgregistered.py).

Of note, I have an implementation for `PgParamDelegate` as well that lets a table cell of a certain type have that `Parameter`'s delegate (i.e. an `int` table field would receive the `int` Parameter spinbox as a delegate). Do you think this is worth considering implementing?

In practice, my table view knows the parameter type and sensible default for each column, and uses them to construct a `Parameter` with an `itemClass` that implements `makeWidget()`. This widget is extracted and used as the table delegate (see an example video of usage below):

https://user-images.githubusercontent.com/23620506/106396329-2da67680-63d5-11eb-8a47-3228c7975a80.mp4

**Comment by ntjess** [comment]
Some parameters are specific to my needs, but I believe the following are general and could apply to an arbitrary audience:
* filepicker
  - existing: bool Whether to pick an existing file ("open") or new file ("save as")
  - asFolder: bool, Whether to select folders or files
  - startDir: string, where to set the default starting directory
  - selectMultiple: bool, if `asFolder=False` this will determine whether multiple files can be selected
  - winTitle: Title of file dialog
* slider
  - limits: [start, stop] numbers
  - step: Defaults to 1, the spacing between each slider tick
  - span: Instead of limits + step, `span` can be set to specify the range of slider options (e.g. `np.linspace(-pi, pi, 100)`)
  - format: format string to determine number of decimals to show, etc. Defaults to display based on span dtype
  - precision: int, number of decimals to keep for float tick spaces
* checklist: Group with clickable boolean children to specify a list of allowed options
  - limits: Name of string options
  - exclusive: bool. If `True`, a single string with the checked option will be returned instead of a list. Checking one box unchecks the previous option. If `False`, any number of child options can be checked. The `value()` of the widget will be a list of each `True` child. See the video below as an example of an `exclusive` checklist
* popuplineeditor: a video shows usage in the previous comment.
  - limits: Options to choose from, much like a dropdown list
* keyseq: Video shows usage in previous comment
  - value: Key sequence, e.g. "Ctrl+C"

https://user-images.githubusercontent.com/23620506/106397073-942d9380-63d9-11eb-8007-0f9f474892cb.mp4

**Comment by j9ac9k** [comment]
Hi @ntjess 

Sorry I haven't posted here; looks like you've implemented some really handy features.  Recently I've had my hands full migrating to github actions, and the head aches involved with that transition.  

A little context.  We have many open issues/PRs with respect to parameter trees (you're likely aware).  Parameter Trees are a feature I've never really used so I haven't felt comfortable evaluating or addressing those.  @lidstrom83 decided he would tackle the bulk of them with #1469 .  Since he started work on that PR, he had some real-life circumstances arise that have resulted in him having to postpone pyqtgraph work for the time being.

Back to this issue, your suggested features look useful!  We would certainly welcome PRs adding to them. Probably should modify the example to highlight those features

Also, the slider widget, we have an open PR (#398) how would your compare to those?  Also looks like #665 implements a slider as well (and some other types as well).

I'm also going to tag @ixjlyons as he has worked to expand documentation for parameter trees and may have more input.

Where I'm currently at right now is I'm considering merging #1469 as it is (as it does quite a few fixes), and then look to tackle other parameter tree related issues... but I would certainly welcome your input on how you think we should merge/address the remaining open issues, _especially_ as you look to make frequent use of this feature.

**Comment by ixjlyons** [comment]
Just a quick comment on the slider while it's on my mind - one of the touted advantages of the model/view design is potentially having different item representations of a parameter. So it seems like the slider would be an alternative graphical interface for the float `SimpleParameter` rather than a new `Parameter` type. I'm not sure what the best way to achieve that is in practice, though, or if it's even worth trying to avoid a 1-to-1 mapping of a `Parameter` type to a `ParameterItem` type.

**Comment by ntjess** [comment]
Thanks for the context.

It looks like the sliders you referenced are pretty similar, but they don't let you specify a `span`, you must specify a `limit` range and optional `step` (I'm not attached to this enough to declare my implementation better, I just know I find it useful in my code base).

#665 also has a file picker parameter, but it's less modular (i.e. you specify the granular file loading function rather than specifying flags for operation like `asFolder`, `existing`, etc. My personal preference is a modular approach, but of course there are reasons to use both.

I'm waiting on submitting a PR until I get some feedback on whether these features are worthwhile before putting time and effort into this.

@ixjlyons Interesting point, this holds for my `popuplineeditor` vs `list` parameter too (if you notice, the options are identical). Do you feel something like a `ui`/`frontend`/etc. flag in `opts` would be more modular?

**Comment by ntjess** [comment]
Slider + file picker now in #1844

**Comment by papapel** [comment]
Hello @ntjess,

It seems you are doing a great work to enhance parameter feature. Thank you.
I designed similar parameter types according to my need. Please find the image below which shows you them:

![image](https://user-images.githubusercontent.com/56733422/135089587-4fa2fc12-102b-408a-b103-77170bd29d60.png)

Some ideas I would like to suggest to you (for inspiration) according the previous picture:
- span action = the button size takes the whole row (it's anecdotal but I find that nicer)
- slider = value can be set from the slider or from the textedit
- hexa = hexadecimal representation of the integer parameter
- '?' button = open a dialog to show the parameter description (please refer to the image below)
![image](https://user-images.githubusercontent.com/56733422/135091302-4b1bfe38-caf5-4b2e-8fc6-0dbc1a3e3c5c.png)
- key-map-slider (not implemented yet) = 
   - the parameter shows a slider to select a value between 0 and 3.3 with 1023 steps
   - the real parameter value is the result of this formula: int(round(value/3.3*1023))


Other thoughts or comments (for discussion):
- I often use pyqtgraph for the parameter tree feature. But I don't need the other features. Is it interesting to make the parameter tree like a standalone python package? It could be easier to maintain and release new versions without waiting for the work on the other features?
- It is very nice to also use parameter tree in non graphical projects. However I meet some limitation with pyqtsignal when I use the tree in pure python threads. The signals are not emitted (specially when a new value for a parameter is set in a thread). I found some workaround. But maybe: Is it interesting to remove the Qt dependency in the core parametertree feature, and only call Qt in the widget parameter type?

**Comment by ntjess** [comment]
Hi @papapel thanks for the suggestions. Here are my thoughts;
* I'm a little confused by the `span` action, it seems like you can just call `tree.header().resizeSection(1, width)` instead of needing a new parameter. Is there a benefit to the approach of having a new parameter for this?
* This seems like a somewhat niche use case, since most of the time you can make `float` and `slider` parameters and tie together their `sigValueChanged` signals. There is an example in the `pyqtgraph.examples` suite showing this.
* For hexadecimals I'd be interested to hear from others (@j9ac9k @ixjlyons) if it's worth meriting a new parameter type. I don't know whether this is considered niche or not
* I'm less inclined to have a `?` button in each row, just because it would add a lot of clutter to the parameter tree where space might already be a concern. If it simply shows the documentation, having 8 `float` boxes would also create 8 `?` buttons with the same help options. Qt has a built-in framework for `whatsThis` events; if you wanted to spawn help text I recommend using that since it is more application agnostic and doesn't require screen real estate.
* The `key-map-slider` functionality you suggested is possible with the slider on `master` -- You can set the `span` option with a list of values to use, i.e. `p = Parameter.create(name='test', type='slider', span=np.linspace(0, 3.3, 1023))`. Does this sound like what you are after?

_____
* I tend to use parameter trees a bunch for my application as well and was thinking about how to possibly separate out the logic. The other advantage would be to alleviate some circular imports in the `pyqtgraph` repository (some imports have to be inside functions because of this dependency relationship). However, this is a bit tricky since some widgets are shared by both the `parametertree` module and the rest of the `pyqtgraph` module (i.e. `SpinBox`, `GradientWidget`). When I have more time I would also like to turn the viewbox context menu options into a parameter tree (there are numerous benefits!), which would further complicate imports if the modules aren't split a bit more. That being said, it probably wouldn't be released as a separate version-controlled repo and pypi package. There are a few other features in `pytgraph` that could be considered standalone, but for the time being it would be difficult to maintain all the separate repos. `pyqtgraph` only depends on `numpy` and some form of `Qt`, which is the same for parameter trees. So there's not much disadvantage to keeping them together, except for (as you noted) the possible release schedule. However, I actually think this is a good thing, since things that sit in the pipeline have the change to be improved, (somewhat) guaranteed to work correctly, and have more eyes on them than if they were totally separate.
  - I _would_ like to eventually be able to use Parameters the same way as regular widgets without a separate import (i.e. `pg.Parameter`), which might(?) be possible after I think hard about the shared widget implementation problem.
* Since all parameters inherit from `QTreeWidgetItem`, it would be hard to break this paradigm and remove `Qt` from the implementation. Signals and slots are used heavily throughout pyqtgraph and would have several advantages compared to other coupling logic. That being said, you should be able to use them in a headless application as long as you make a `QApplication.instance` (@j9ac9k feel free to correct me on this). For instance, I know I've used the `minimal` platform and signals/slots were working fine for me.

Again, @papapel thanks for thinking through this and offering suggestions, hopefully this answered most of your questions. I'm happy to keep the discussion going!

**Comment by j9ac9k** [comment]
> * Since all parameters inherit from `QTreeWidgetItem`, it would be hard to break this paradigm and remove `Qt` from the implementation. Signals and slots are used heavily throughout pyqtgraph and would have several advantages compared to other coupling logic. That being said, you should be able to use them in a headless application as long as you make a `QApplication.instance` (@j9ac9k feel free to correct me on this). For instance, I know I've used the `minimal` platform and signals/slots were working fine for me.

You can most certainly use pyqtgraph (or Qt in general) in headless applications.  If we couldn't, our CI test suite would be in major trouble.

Here you can see where in our CI suite you can see what packages we install (on linux) to make the headless systems work:

https://github.com/pyqtgraph/pyqtgraph/blob/master/.github/workflows/main.yml#L71-L78

There are some other OS specific gotchas that you may periodically run into.

**Comment by papapel** [comment]
Thank you for answering me.

- `span action`: there is no benefit to have a new parameter, an option instead? But this is not very important.
- `float/int-and-slider` Actually I like when you have both ways to modify the parameter value. Like on the parameter tree of Adobe After Effects (see below).
- `hexa`: This is actually an integer parameter but I modified the spinbox validator to display hexadecimal character only. I use this kind of parameter to have a mirror of the register map of my embedded device. And sub-registers which are parts of the register can be `int`, `float` or `bool` parameters. Please find the screenshot below:
![image](https://user-images.githubusercontent.com/56733422/135154942-e9676abe-c683-4754-8c22-e2221d83d1fa.png)
I create a specific parameter and I didn't add an option because I need to set the number of bits of my hexadecimal parameter (16, 32 or 64bits). And by the way, I added a parameter filter feature as well as collapse/expand button which can be very nice when we have a huge parameter tree.
- `?` button: Actually, I set a specific description for each parameter: a float parameter can be used to set a frequency or a voltage level and I explain in the description the effect of this parameter, the associated feature, the range etc. I use an option to set the path of the file where I wrote the description. And the `?` button is hidden when no file is set. The `whatsThis` can meet the need but I chose to add a `?` button because I would like the user knows quickly which parameter has a description. Because I generally set the description for a group of parameters and not for each parameter.
- `key-map-slider`: Ok, this is a very specific need. I would like to do a mix between a `dict parameter` and the `span` option of the `slider` parameter. Eg: my slider displays a range between 0.0V and 3.3V. When the user sets 0.0V (the min), the parameter value is actually 0, and when the user sets 3.3V (the max), the parameter value is actually 1023. Currently, I connect a callback to the valueChanged signal, I take the parameter value and the callback performs the conversion from 3.3 to 1023. But I have a lot of parameter like that. It could be nice to me to handle automatically this case.
#
I find that the parameter tree on Adobe After Effects is very nice.
**int/float and slider parameter**
![image](https://user-images.githubusercontent.com/56733422/135152085-804aa830-ed54-43b4-8123-7974ade2c251.png)
**parameter examples**
![image](https://user-images.githubusercontent.com/56733422/135157959-4c701014-6934-487a-bc16-ffec82d4595e.png)
some ideas:
- an option to display the full range of the slider or display the slider between 0% and 100%
- a circular parameter (useful when you want to set a phase)
#
- It makes sense. Thank you for your answer.
- @ntjess, @j9ac9k thank you for your answers. I will check that.

Obviously, all I am sharing you is suggestion only. I just want to say : if I had this need, maybe other ones could have it as well. If not, I will continue to create my own parameters and register them (and it works well! ) :slightly_smiling_face:
